### PR TITLE
Reverse dependency on components

### PR DESCRIPTION
**Pull request by waralex**
DashBase.jl package added. Dash depends on DashBase and core components, core components depends on DashBase and other components depends on Dash. 
For testing until the corresponding packages are added to the Registrator you need to install the packages in this order:
```julia
Pkg.add(PackageSpec(url="https://github.com/waralex/DashBase.jl.git"))
Pkg.add(PackageSpec(url="https://github.com/waralex/dash-html-components.git", rev="jl_reverse_test"))
Pkg.add(PackageSpec(url="https://github.com/waralex/dash-core-components.git", rev="jl_reverse_test"))
Pkg.add(PackageSpec(url="https://github.com/waralex/dash-table.git", rev="jl_reverse_test"))
Pkg.add(PackageSpec(url="https://github.com/plotly/Dash.jl.git", rev="dev"))
```
Corresponding PR to the generator https://github.com/plotly/dash/pull/1302

If this option is accepted, then I will need to move the DashBase to the plotly repository. You will also need to update the `jl` branches for the main components

### CODE REVIEW

**Comment by alexcjohnson** [comment]
> For testing until the corresponding packages are added to the Registrator you need to install the packages in this order

To confirm though, once they are added to the registry this will be just:
```julia
Pkg.add("dash")
```
And the other 4 come along automatically, correct? If so, this looks perfect!

File: .circleci/config.yml [hunk 1]
```diff
@@ -28,7 +28,7 @@ jobs:
...
+            julia -e 'using Pkg; Pkg.update(); Pkg.add(PackageSpec(url="https://github.com/waralex/DashBase.jl.git"));  Pkg.add(PackageSpec(url="https://github.com/waralex/dash-html-components.git", rev="jl_reverse_test")); Pkg.add(PackageSpec(url="https://github.com/waralex/dash-core-components.git", rev="jl_reverse_test"));Pkg.add(PackageSpec(url="https://github.com/waralex/dash-table.git", rev="jl_reverse_test")); Pkg.add(PackageSpec(url="https://github.com/plotly/Dash.jl.git", rev=ENV["CIRCLE_BRANCH"])); Pkg.build("Dash");Pkg.test("Dash", coverage=true); function precompile_pkgs(); for pkg in collect(keys(Pkg.installed())); if !isdefined(Symbol(pkg), :Symbol) && pkg != "Compat.jl"; @info("Importing $(pkg)..."); try (@eval import $(Symbol(pkg))) catch end; end; end; end; precompile_pkgs()'
```
Comment (alexcjohnson) [inline .circleci/config.yml:31]: It's awfully hard to tell what's changing here - perhaps we can move this to a separate `.jl` file? To make clear that it's meant only for CI we can put it in the `.circleci` folder - similar to what we do with the [plotly.js test script](https://github.com/plotly/plotly.js/blob/fbb5742241dcc46b84f508bad812c975d7f06b16/.circleci/config.yml#L50).

**Comment by waralex** [comment]
@alexcjohnson 

>And the other 4 come along automatically, correct? If so, this looks perfect!

The user still needs to Pkg.add component packages if they plan to use them in their environment. In his entourage is added to the component version that satisfies the dependencies Dash. I.e. in Julia is 2 different actions - physical installation package and add a package to the current environment. Pkg.add(dash) physically install the component packages, but the user still needs to add them to their environment. The user does not need to add DashBase to their environment.

**Comment by alexcjohnson** [comment]
As discussed on Slack, let's proceed with this. Aside from my comment about the unit test CI script this looks in good shape to me!

The only potential change I can think of after this is bundling the rest of the core as sub-packages of `Dash` - `Dash.HtmlComponents` instead of `DashHtmlComponents`. We don’t have to decide that now anyway, we can always do it later by importing and re-exporting them inside `Dash`. If we did that, the package structure would be different from Python and R, but we’d be back to a single-item installation. Also might be tricky to force unused sub-packages not to load their javascript, though for the core packages that’s not a big deal anymore since they all support async loading - the initial stubs are small.

**Comment by waralex** [comment]
@alexcjohnson Ok. I will move `DashBase` to the Plotly repository. On your part, you need to finish the generator (https://github.com/plotly/dash/tree/1189-julia-components) and update the `jl` branches of the components so that I can specify them in the README for the dev branch.

**Comment by alexcjohnson** [comment]
Is this PR obsolete now that we have #54?

**Comment by waralex** [comment]
>Is this PR obsolete now that we have #54?

Yes, I'll close it

**Pull request closed without merging**